{"cells":[{"cell_type":"markdown","metadata":{"id":"l3m0STYsnkYm"},"source":["#### 【NGS解析】Arabidopsis RNA-Seq データの DGE解析に向けた前処理   <br>&nbsp; RNAseq_01  : Stepwise Pipeline"]},{"cell_type":"markdown","metadata":{"id":"JbV9AAsknkYn"},"source":["--------------------------------------------------------------------------</br>\n","\n","  このノートブックはGoogle Colab上で動かすことを想定しており、  <br>\n","また、１セルずつ実行して１処理ずつ確認しながら進めていく形になっています。  \n","\n","✅Arabidopsis thaliana の<strong>single-end</strong> RNA-Seq データを使用した、  <br>&nbsp;&nbsp; &nbsp;&nbsp;DGE解析に向けたデータの前処理です。  \n","✅Google Drive フォルダに次のファイルが入れてある前提です。\n","* drive_path :   accession_list.txt, param1~4.txt\n","* common_ref :   araport11.gtf, TAIR10_chr_all.fas, smr_v4.3_default_db.fasta  \n","<small>💡バックスラッシュ＼が円マーク￥で表示される場合はブラウザのフォントを英字フォントに変更する</small>  \n","\n","※詳細はReadmeに記載しています。  \n","--------------------------------------------------------------------------</br>\n"]},{"cell_type":"markdown","metadata":{"id":"AXb3EK14rEY1"},"source":["🛠️ 初期設定：Google Drive をマウント（セッション再接続ごとに実行）"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"A87SFlNlotuL"},"outputs":[],"source":["from google.colab import drive\n","drive.mount(\"/content/drive/\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"GlUKIjdoHtCo"},"outputs":[],"source":["drive_path = \"/content/drive/MyDrive/ngs_analysis/rna-seq_01\"  # データを退避する先のパスをpython変数で定義\n","common_ref = \"/content/drive/MyDrive/ngs_analysis/common_ref\"    #  参照ファイルが保存してあるパスをpython変数で定義"]},{"cell_type":"markdown","metadata":{"id":"PnqwQKRK9ti9"},"source":[" 🛠️ 初期設定：Micromambaをインストール（セッション再接続ごとに実行）"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"spa_4rR7yNhz"},"outputs":[],"source":["!curl -Ls https://micro.mamba.pm/api/micromamba/linux-64/latest | tar -xvj -C /root/ --strip-components=1"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"9K0frutFb6D-"},"outputs":[],"source":["import os"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wuXJm89rcAxX"},"outputs":[],"source":["os.environ[\"PATH\"] = os.path.abspath(\"/root/\") + os.pathsep  + os.environ[\"PATH\"] # 環境変数に/root/パスを追加\n","print(os.environ[\"PATH\"])"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7J6QYSJAcGft"},"outputs":[],"source":["!micromamba create -y -n ngs  # NGS用の仮想環境を作成"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"gg7Lz2KYcsz9"},"outputs":[],"source":["os.environ[\"PATH\"] = os.path.abspath(\"/root/.local/share/mamba/envs/ngs/bin\") + os.pathsep  + os.environ[\"PATH\"]  # 環境変数にパスを追加\n","print(os.environ[\"PATH\"])"]},{"cell_type":"markdown","metadata":{"id":"38HQAffbveTN"},"source":["---\n","⬇️ 処理ここから ⬇️"]},{"cell_type":"markdown","metadata":{"id":"YOS4eteK98gE"},"source":["■ 解析対象のリードデータを SRA 互換リポジトリから取得してFASTQ 形式に変換"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"TuUJ6Y12W3N3"},"outputs":[],"source":["!mkdir sra_data"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"tziTaRMpgyPg"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda  sra-tools"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"J5sp5b-QpE-O"},"outputs":[],"source":["!prefetch --option-file {drive_path}/accession_list.txt  -O /content/sra_data/"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"yRho1mTTsFwH"},"outputs":[],"source":["!fasterq-dump  /content/sra_data/*  -O /content/"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"AEXMk7kGoPTF"},"outputs":[],"source":["!cp /content/*fastq  {drive_path}  # fastqファイルをGoogle Driveに退避"]},{"cell_type":"markdown","metadata":{"id":"K-4YNKsa-VmA"},"source":["■ fastqデータの品質をチェック 【FastQC】"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"OfwrAwyvqORZ"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda fastqc"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"twGQ07iGqqE4"},"outputs":[],"source":["!fastqc /content/*.fastq"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bkDQDblmKN8q"},"outputs":[],"source":["!cp /content/*.html  {drive_path}  # FastQC ReportをGoogle Driveに退避"]},{"cell_type":"markdown","metadata":{"id":"QYJFykmHouuY"},"source":["---\n","⏸️ セッションの一時終了  <br>\n","<small>&nbsp;&nbsp; &nbsp;&nbsp;※接続オプションから「接続を解除して削除」を選択して現在のセッションを終了  \n","\n","  ・Google Drive に保存された.htmlファイル（FastQC Report）をダウンロードし、ブラウザで開いて確認します。  <br>\n","  ・FastQC の結果を参考にしてトリミング方法を検討します。  <br></small>\n"]},{"cell_type":"markdown","metadata":{"id":"gvZ1XbKBCQJo"},"source":["\n","&nbsp;&nbsp; &nbsp;&nbsp;<small>💡 セッション終了後すぐに再接続すると、リソースの解放処理が完了しておらず、環境に不整合が生じることがあるため、数分時間をあける。</small>"]},{"cell_type":"markdown","metadata":{"id":"1Fd0720aj4JZ"},"source":["▶️ セッションの再開  <br>\n","<small> &nbsp;&nbsp; &nbsp;&nbsp;※接続オプションから「ホストされているランタイムに接続」を選択してセッションを再接続  </small><br>\n","\n","<small>&nbsp;&nbsp;🎯以下のステップを再実行：  <br>\n","&nbsp;&nbsp; &nbsp;&nbsp;🛠️初期設定：Google Drive をマウント  <br>\n","&nbsp;&nbsp; &nbsp;&nbsp;🛠️初期設定：Micromambaをインストール</small>"]},{"cell_type":"markdown","metadata":{"id":"G88zNjlg-ge2"},"source":["■ リード配列をカスタムトリミング【cutadapt】&【sortMeRNA】"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"glfDSt8EE2Zw"},"outputs":[],"source":["!cp {drive_path}/*.fastq  /content/  #退避していたfastqファイルをセッションにコピー"]},{"cell_type":"markdown","metadata":{"id":"VKH0618TJ8Gy"},"source":["ー cutadaptを用いた人工配列と低品質部位のトリミング"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"egUSkgmfJdxL"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda cutadapt"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7-H77K3KQGpH"},"outputs":[],"source":["!cp {drive_path}/params*.txt /content/  # cutadapt用のオプションと引数を指定したファイルをセッションにコピー"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"F1AcjXJ9go6o"},"outputs":[],"source":["%%bash\n","PARAMS1=$(<params1.txt)  # for removing structurally invalid reads\n","PARAMS2=$(<params2.txt)  # for trimming Template Switching Oligo sequences\n","PARAMS3=$(<params3.txt)  # for trimming Nextera adapters and primer sequences\n","PARAMS4=$(<params4.txt)  # for quality filtering and tail trimming\n","                                              # used for fastq data from SmartSeq2+Nextera chemistry, NextSeq\n","for file in /content/*.fastq; do\n","     cutadapt   $PARAMS1  -o temp1  \"$file\"   > \"${file%.fastq}_report1.txt\"\n","     cutadapt   $PARAMS2  -o temp2   temp1 > \"${file%.fastq}_report2.txt\"\n","     cutadapt   $PARAMS3  -o temp3   temp2 > \"${file%.fastq}_report3.txt\"\n","     cutadapt   $PARAMS4  -o \"${file%.fastq}_cut\"  temp3 > \"${file%.fastq}_report4.txt\"\n","     rm temp1 temp2 temp3;\n","done\n"," # 💡 左のファイル一覧で出力ファイルが増えていくことで進捗を確認"]},{"cell_type":"markdown","metadata":{"id":"Ko83oUneJdQD"},"source":["ー sortMeRNAを用いたrRNAの除去処理"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"2XquqJKuviob"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda sortmerna=4.3.7"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"HoLaCBIVwSKw"},"outputs":[],"source":["!cp {common_ref}/smr_v4.3_default_db.fasta  /content/"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"48W_243y0uz6"},"outputs":[],"source":["!for file in /content/*_cut; do sortmerna -ref smr_v4.3_default_db.fasta -reads \"$file\" -workdir /content/smr_work/ -no-best -a 4 -fastx -other \"${file%_cut}_cleaned\" &&  rm /content/smr_work/kvdb/*; done"]},{"cell_type":"markdown","metadata":{"id":"YOL-rGqQ-rMA"},"source":["■ トリミング後のfastq データの品質をチェック 【FastQC】"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"rI5-nnYj1m26"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda fastqc"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"q6WROg4O9Dg8"},"outputs":[],"source":["!fastqc /content/*_cleaned.fq"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"DE9Iq9WY4vxk"},"outputs":[],"source":["!cp /content/*_cleaned.fq {drive_path}  # トリミング後のfastqファイル(拡張子無し)をGoogle Driveに退避\n","!cp /content/*.html {drive_path}         # FstQC ReportをGoogle Drive に退避"]},{"cell_type":"markdown","metadata":{"id":"WQ6nhCaKa11M"},"source":["⏸️ セッションの一時終了  <br>\n","       &nbsp;&nbsp;  &nbsp;&nbsp; 結果の確認<br>\n","▶️ セッションの再開"]},{"cell_type":"markdown","metadata":{"id":"9OpKGphfAqvh"},"source":["■ リード配列を参照ゲノムにマッピング 【HISAT2】"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jp3GNyBiKXUK"},"outputs":[],"source":["!cp  {drive_path}/*_cleaned.fq  /content/  #保存していたトリミング後fastqファイルをセッションにコピー"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"S0q4b-U6J4bK"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda hisat2"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"FMKmZYn1m0j7"},"outputs":[],"source":["!hisat2_extract_splice_sites.py {common_ref}/araport11.gtf > splice_sites.txt  #\tスプライスサイト情報を抽出"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"RI2eea9Zsl4g"},"outputs":[],"source":["!hisat2_extract_exons.py {common_ref}/araport11.gtf > exons.txt  # エクソン情報を抽出"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"RhkvU1MOtHjh"},"outputs":[],"source":["!hisat2-build --ss splice_sites.txt --exon exons.txt   {common_ref}/TAIR10_chr_all.fas  at_index #参照ゲノムをもとにhisat2のインデックスファイルを作成"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"UXyP2uF_FL87"},"outputs":[],"source":["!for file in *_cleaned.fq ; do hisat2 -x at_index -p 2  --phred33  -q -U \"$file\" -S \"${file%_cleaned}_aligned.sam\";done"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"iEjoaeLv0zNT"},"outputs":[],"source":["!cp /content/*.ht2 {common_ref}  #  HISAT2インデックス（splice site情報・exon情報含む）をGoogle Driveに保存"]},{"cell_type":"markdown","metadata":{"id":"KXyUfWmZDD3L"},"source":["■ SAM→BAM変換 (テキストからバイナリへ) 【Samtools】\n","  "]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"sbZ2sPHjfoK8"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda samtools"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"KWTByRy68vIT"},"outputs":[],"source":["!for file in /content/*_aligned.sam; do samtools view -b \"$file\" -o \"${file%.sam}.bam\"; done  # SAM→BAM変換"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"3CrMUBlh7IMB"},"outputs":[],"source":["!for file in /content/*.bam; do samtools sort \"$file\" -o \"${file%_aligned.bam}_sorted.bam\"; done # ソート"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"pTmOpjg243qv"},"outputs":[],"source":["!for file in /content/*_sorted.bam; do samtools index \"$file\"; done  # BAMインデックスファイルを作成（Genome Viewerで使用）"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ObRgdYO1jCq2"},"outputs":[],"source":["!cp /content/*sorted.bam  {drive_path}  # ソート済みBAMファイルをGoogle Driveに保存"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"8xQuwUDAY1u2"},"outputs":[],"source":["!cp /content/*sorted.bam.bai  {drive_path}  # BAMインデックスファイルをGoogle Driveに保存"]},{"cell_type":"markdown","metadata":{"id":"lCzp4mJ1SCkw"},"source":["■ マッピング結果に基づきリード数を定量化 【subread】"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"rgXipxU7Aeuz"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda subread"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"he--sTGhAiu5"},"outputs":[],"source":["!featureCounts -T 2  -t exon -g gene_id  -a {common_ref}/araport11.gtf -o counts.txt  /content/*_sorted.bam"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"MApV-2akYuok"},"outputs":[],"source":["!cp /content/counts.txt  {drive_path}  # リードカウントファイルをGoogle Driveに保存"]},{"cell_type":"markdown","metadata":{"id":"AVDYQgFk2yhW"},"source":["🔚このNotebookはここまで  <br>\n","&nbsp;&nbsp; &nbsp;&nbsp;作成したカウントマトリックス（counts.txt）を用いてDGE解析に進みます。  \n","\n","※ rna-seq_02 および rna-seq_03 では、ファイルを一時保存せず、パイプ入力によるストリーミングで逐次処理を行っています。\n","\n","\n","\n","---\n","\n"]}],"metadata":{"colab":{"provenance":[{"file_id":"https://github.com/takany-bioinfo/rna-seq-learning/blob/main/rna-seq_01/rna-seq_01.ipynb","timestamp":1759288060272}]},"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.7"}},"nbformat":4,"nbformat_minor":0}