{"cells":[{"cell_type":"markdown","metadata":{"id":"l3m0STYsnkYm"},"source":["###  NGS: *Arabidopsis thalian*a RNA-Seq Preprocessing for DGE Analysis<br>  \n"," &nbsp;  rna-seq_03  : Streamlined Pipeline, **Paired-end**"]},{"cell_type":"markdown","metadata":{"id":"JbV9AAsknkYn"},"source":["---\n","\n","This notebook is designed for performing NGS analysis on Google Colab.\n","\n","*   Preprocessing Arabidopsis thaliana paired-end RNA-Seq data for subsequent differential gene expression (DGE) analysis.<br>\n","\n","*   The following files are expected to be located in a specific folder on Google Drive:<br>\n","accession_list.txt, at_index*.ht2, araport11.gtf<br>  \n","\n","Note: See the README for more details.<br>\n","\n","---\n"]},{"cell_type":"markdown","metadata":{"id":"AXb3EK14rEY1"},"source":["Mounting Google Drive"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"A87SFlNlotuL"},"outputs":[],"source":["from google.colab import drive\n","drive.mount(\"/content/drive/\")"]},{"cell_type":"markdown","metadata":{"id":"PnqwQKRK9ti9"},"source":["Installing micromamba"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"spa_4rR7yNhz","collapsed":true},"outputs":[],"source":["!curl -Ls https://micro.mamba.pm/api/micromamba/linux-64/latest | tar -xvj -C /root/ --strip-components=1"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"9K0frutFb6D-"},"outputs":[],"source":["import os\n","os.environ[\"PATH\"] = os.path.abspath(\"/root\") + os.pathsep + os.environ[\"PATH\"] # Add /root/ to the environment variables"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7J6QYSJAcGft"},"outputs":[],"source":["!micromamba create -y -n ngs  # Create virtual environment for NGS"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"gg7Lz2KYcsz9"},"outputs":[],"source":["os.environ[\"PATH\"] = os.path.abspath(\"/root/.local/share/mamba/envs/ngs/bin\") + os.pathsep + os.environ[\"PATH\"]\n","# Add env path to the environment variables"]},{"cell_type":"markdown","metadata":{"id":"38HQAffbveTN"},"source":["Installing Required Tools"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"tziTaRMpgyPg"},"outputs":[],"source":["!micromamba install -y -q -n ngs -c conda-forge -c bioconda  sra-tools fastqc fastp hisat2 samtools subread"]},{"cell_type":"markdown","metadata":{"id":"99y1cTV3Fs4A"},"source":["Raw Reads Processing: Downloading, Trimming, Mapping, and Binary Conversion"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"89TRQIr-Z0J3","collapsed":true},"outputs":[],"source":["%%bash\n","set -e\n","set -o pipefail\n","now=$(date '+%Y%m%d_%H%M%S')\n","logfile=\"/content/run_${now}.log\"\n","exec > >(tee -a \"$logfile\") 2>&1\n","\n","drive_path=\"/content/drive/MyDrive/ngs_analysis/rna-seq_03\"\n","comon_ref=\"/content/drive/MyDrive/ngs_analysis/common_ref\"\n","\n","while read file; do\n","    echo \"------------------------------------------------------------\"\n","    start_time=$(date +%s)\n","    echo \"<<< Start processing for ${file} :  [`date '+%F %T'`] >>> \"\n","\n","    echo \"### [PREFETCH ] : ${file} ###\"\n","    prefetch \"${file}\" -O /content/\n","\n","    echo \"### [FASTQ-DUMP] : ${file} ###\"\n","    fasterq-dump /content/\"${file}\" --split-files -O /content/\n","    rm  /content/\"${file}\"/\"${file}.sra\"  # Remove sra files\n","\n","    echo \"### [ FASTP ] : ${file} ###\"\n","    fastp   --in1 /content/\"${file}_1.fastq\" \\\n","              --in2 /content/\"${file}_2.fastq\" \\\n","             --out1 /content/\"${file}_1_cleaned.fastq\" \\\n","             --out2 /content/\"${file}_2_cleaned.fastq\" \\\n","             --html /content/\"${file}.html\"\n","    cp /content/\"${file}.html\"  ${drive_path}   # Save fastp report file to Google Drive\n","    rm  /content/\"${file}_1.fastq\"   /content/\"${file}_2.fastq\" # Remove fastp files\n","\n","    echo \"### [ HISAT2 ] : ${file} ,  [SAMâ†’BAM] ###\"\n","    hisat2 -x ${comon_ref}/at_index \\\n","           -1 /content/\"${file}_1_cleaned.fastq\" \\\n","           -2 /content/\"${file}_2_cleaned.fastq\" \\\n","           -p 4 \\\n","           -k 2 \\\n","           --phred33 |\n","           samtools view -@ 4 -b  -o /content/\"${file}.bam\"\n","\n","    rm  /content/\"${file}_1_cleaned.fastq\"   /content/\"${file}_2_cleaned.fastq\"  # Remove cleaned fastq files\n","\n","     echo \"<< Completed processing for ${file} >>\"\n","     end_time=$(date +%s)\n","     elapsed=$((end_time - start_time))\n","     echo \"Time for ${file}: ${elapsed} seconds\"\n","     echo \"------------------------------------------------------------\"\n","     cp /content/*.log ${drive_path} # Save log file to Google Drive\n"," done < ${drive_path}/accession_list.txt\n","\n"," echo \"### [featureCounts] ###\"\n"," featureCounts  -T 4  -p \\\n","            -t exon \\\n","            -g gene_id \\\n","            -a ${comon_ref}/araport11.gtf \\\n","            -o counts.txt \\\n","              /content/*.bam\n","cp /content/counts.txt  ${drive_path}   # Save read count file to Google Drive\n","cp /content/counts.txt.summary  ${drive_path}\n","\n","echo \"<<< featureCounts completed - result saved : [`date '+%F %T'`] >>>\"\n","cp /content/*.log ${drive_path} # Save log file to Google Drive\n"]},{"cell_type":"markdown","metadata":{"id":"AVDYQgFk2yhW"},"source":["ðŸ”š End of this notebook:<br>&nbsp;&nbsp; &nbsp;&nbsp; Next, proceed to DEG analysis using the generated count matrix (counts.txt).\n","\n","\n","\n","---\n","\n"]},{"cell_type":"markdown","metadata":{"id":"1Z-EV78fj_V-"},"source":[]}],"metadata":{"colab":{"provenance":[{"file_id":"1SmYECMtpPxLhJySmjPrvfnOqZSxcyJ21","timestamp":1755159157644},{"file_id":"1h85DoxKG5hHQLPltJsIM1Uw38aA_FnyL","timestamp":1744353873890},{"file_id":"1NOOG-t_D9kQX55lEbQ7VJvUa8-RnWOqE","timestamp":1744350150641}]},"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.7"}},"nbformat":4,"nbformat_minor":0}
