{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "l3m0STYsnkYm"
      },
      "source": [
        "#### ã€NGSè§£æã€‘Arabidopsis RNA-Seq ãƒ‡ãƒ¼ã‚¿ã® DGEè§£æã«å‘ã‘ãŸå‰å‡¦ç†   <br>&nbsp; RNAseq_02  : Partially Streamlined Pipeline"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JbV9AAsknkYn"
      },
      "source": [
        "---\n",
        "<small>ã“ã®ãƒãƒ¼ãƒˆãƒ–ãƒƒã‚¯ã¯Google Colabä¸Šã§å‹•ã‹ã™ã“ã¨ã‚’æƒ³å®šã—ã¦ã„ã¾ã™ã€‚  \n",
        "ãƒ»Arabidopsis thaliana ã®single-end RNA-Seq ãƒ‡ãƒ¼ã‚¿ã‚’ä½¿ç”¨ã—ãŸã€DGEè§£æã«å‘ã‘ãŸãƒ‡ãƒ¼ã‚¿ã®å‰å‡¦ç†ã§ã™ã€‚  \n",
        "ãƒ»Google Drive ãƒ•ã‚©ãƒ«ãƒ€ã«æ¬¡ã®ãƒ•ã‚¡ã‚¤ãƒ«ãŒå…¥ã‚Œã¦ã‚ã‚‹å‰æã§ã™ã€‚    \n",
        "&nbsp;&nbsp; &nbsp;&nbsp;accession_list.txt, at_index*.ht2 , araport11.gtf  \n",
        "â€»è©³ç´°ã¯Readmeã«è¨˜è¼‰ã—ã¦ã„ã¾ã™ã€‚  </small>\n",
        "\n",
        "---\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "â–  åˆæœŸè¨­å®šï¼šGoogle Drive ã‚’ãƒã‚¦ãƒ³ãƒˆ"
      ],
      "metadata": {
        "id": "AXb3EK14rEY1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount(\"/content/drive/\")"
      ],
      "metadata": {
        "id": "A87SFlNlotuL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PnqwQKRK9ti9"
      },
      "source": [
        "â–  åˆæœŸè¨­å®šï¼šMicromambaã‚’ã‚¤ãƒ³ã‚¹ãƒˆãƒ¼ãƒ«"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!curl -Ls https://micro.mamba.pm/api/micromamba/linux-64/latest | tar -xvj -C /root/ --strip-components=1"
      ],
      "metadata": {
        "id": "spa_4rR7yNhz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "os.environ[\"PATH\"] = os.path.abspath(\"/root\") + os.pathsep + os.environ[\"PATH\"] # ç’°å¢ƒå¤‰æ•°ã«/root/ã‚’è¿½åŠ "
      ],
      "metadata": {
        "id": "9K0frutFb6D-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!micromamba create -y -n ngs  # NGSç”¨ã®ä»®æƒ³ç’°å¢ƒã‚’ä½œæˆ"
      ],
      "metadata": {
        "id": "7J6QYSJAcGft"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "os.environ[\"PATH\"] = os.path.abspath(\"/root/.local/share/mamba/envs/ngs/bin\") + os.pathsep + os.environ[\"PATH\"]  # ç’°å¢ƒå¤‰æ•°ã«ãƒ‘ã‚¹ã‚’è¿½åŠ "
      ],
      "metadata": {
        "id": "gg7Lz2KYcsz9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!micromamba install -y -q -n ngs -c conda-forge -c bioconda  sra-tools fastqc fastp hisat2 samtools subread"
      ],
      "metadata": {
        "id": "tziTaRMpgyPg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "â–   SRAå–å¾— â†’ ãƒ€ãƒ³ãƒ— â†’ ãƒˆãƒªãƒŸãƒ³ã‚° â†’ ãƒãƒƒãƒ”ãƒ³ã‚° â†’ ãƒã‚¤ãƒŠãƒªå¤‰æ› >>> å®šé‡åŒ–  \n",
        "ã€€ prefetch â†’ fasterq-dump â†’ fastp â†’ hisat2 â†’ samtools view >>> featureCounts\n"
      ],
      "metadata": {
        "id": "99y1cTV3Fs4A"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "set -e\n",
        "set -o pipefail\n",
        "\n",
        "drive_path=\"/content/drive/MyDrive/ngs_analysis/rna-seq_02\"\n",
        "common_ref=\"/content/drive/MyDrive/ngs_analysis/common_ref\"\n",
        "\n",
        "while read file; do\n",
        "    echo \"------------------------------------------------------------\"\n",
        "    echo \"<<< Start processing for ${file} >>> \"\n",
        "\n",
        "    echo \"### [PREFETCH ] : ${file} ###\"\n",
        "    prefetch \"${file}\" -O /content/\n",
        "\n",
        "    echo \"### [FASTQ-DUMP] : ${file} ###\"\n",
        "    fasterq-dump /content/\"${file}\" -O /content/\n",
        "\n",
        "    echo \"### [ FASTQC ] :  ${file} ###\"\n",
        "    fastqc /content/\"${file}.fastq\"\n",
        "\n",
        "    echo \"### [ FASTP ] : ${file} ###\"\n",
        "    fastp -i /content/\"${file}.fastq\" \\\n",
        "            -o /content/\"${file}_cleaned.fastq\" \\\n",
        "            --trim_poly_g \\\n",
        "            --trim_poly_x \\\n",
        "            --length_required 20 \\\n",
        "            -h /content/\"${file}.html\"\n",
        "\n",
        "    echo \"### [ HISAT2 ] : ${file} ,  [SAMâ†’BAMâ†’ sorting] ###\"\n",
        "    hisat2 -x ${common_ref}/at_index \\\n",
        "           -U /content/\"${file}_cleaned.fastq\" \\\n",
        "           -p 2 \\\n",
        "           --phred33 | \\\n",
        "     samtools view -b  | \\\n",
        "     samtools sort -@ 2  -o /content/\"${file}.bam\"   # for genome browser visualization\n",
        "\n",
        "\n",
        "     echo \"### [BAM Indexing] : ${file} ###\"\n",
        "     samtools index /content/\"${file}.bam\"  # for genome browser visualization\n",
        "\n",
        "     echo \"<<< Completed processing for ${file} >>>\"\n",
        "     echo \"------------------------------------------------------------\"\n",
        "\n",
        "done < ${drive_path}/accession_list.txt\n",
        "\n",
        "echo \"### [featureCounts] ###\"\n",
        "featureCounts  -T 2 \\\n",
        "            -t exon \\\n",
        "            -g gene_id \\\n",
        "            -a ${common_ref}/araport11.gtf \\\n",
        "            -o counts.txt \\\n",
        "              /content/*.bam\n",
        "\n",
        "cp /content/counts.txt  ${drive_path}   # Save read count file to Google Drive\n",
        "cp /content/counts.txt.summary  ${drive_path}\n",
        "\n",
        "echo \"### Count file saved ###\"\n",
        "\n",
        "cp /content/*.bai ${drive_path}   # Copy BAM indices if genome browser visualization is intended\n",
        "cp /content/*.bam ${drive_path} # Copy BAM files  if genome browser visualization is intended\n",
        "cp /content/*.html ${drive_path} # Copy FastQC and Fastp reports to assess FASTQ quality before and after trimming\n"
      ],
      "metadata": {
        "id": "89TRQIr-Z0J3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "ğŸ”šã“ã®Notebookã¯ã“ã“ã¾ã§  <br>\n",
        "&nbsp;&nbsp; &nbsp;&nbsp;ä½œæˆã—ãŸã‚«ã‚¦ãƒ³ãƒˆãƒãƒˆãƒªãƒƒã‚¯ã‚¹ï¼ˆcounts.txtï¼‰ã‚’ç”¨ã„ã¦DGEè§£æã«é€²ã¿ã¾ã™ã€‚  \n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n"
      ],
      "metadata": {
        "id": "AVDYQgFk2yhW"
      }
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.7"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}